activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb

```diff
@@ -6,50 +6,7 @@ module SQLite3
       module SchemaStatements # :nodoc:
         # Returns an array of indexes for the given table.
         def indexes(table_name)
-          exec_query("PRAGMA index_list(#{quote_table_name(table_name)})", "SCHEMA").map do |row|
-            # Indexes SQLite creates implicitly for internal use start with "sqlite_".
-            # See https://www.sqlite.org/fileformat2.html#intschema
-            next if row["name"].starts_with?("sqlite_")
-
-            index_sql = query_value(<<~SQL, "SCHEMA")
-              SELECT sql
-              FROM sqlite_master
-              WHERE name = #{quote(row['name'])} AND type = 'index'
-              UNION ALL
-              SELECT sql
-              FROM sqlite_temp_master
-              WHERE name = #{quote(row['name'])} AND type = 'index'
-            SQL
-
-            /\bON\b\s*"?(\w+?)"?\s*\((?<expressions>.+?)\)(?:\s*WHERE\b\s*(?<where>.+))?\z/i =~ index_sql
-
-            columns = exec_query("PRAGMA index_info(#{quote(row['name'])})", "SCHEMA").map do |col|
-              col["name"]
-            end
-
-            orders = {}
-
-            if columns.any?(&:nil?) # index created with an expression
-              columns = expressions
-            else
-              # Add info on sort order for columns (only desc order is explicitly specified,
-              # asc is the default)
-              if index_sql # index_sql can be null in case of primary key indexes
-                index_sql.scan(/"(\w+)" DESC/).flatten.each { |order_column|
-                  orders[order_column] = :desc
-                }
-              end
-            end
-
-            IndexDefinition.new(
-              table_name,
-              row["name"],
-              row["unique"] != 0,
-              columns,
-              where: where,
-              orders: orders
-            )
-          end.compact
+          new_indexes_from_fields(index_definitions(table_name), table_name)
```

Comment:
Why don't you put `new_indexes_from_fields` just below `indexes`?
It reduces the diff dramatically.

```diff
diff --git a/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb b/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb
index e64e995e1a..a8ff4f01db 100644
--- a/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb
+++ b/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb
@@ -6,7 +6,11 @@ module SQLite3
       module SchemaStatements # :nodoc:
         # Returns an array of indexes for the given table.
         def indexes(table_name)
-          exec_query("PRAGMA index_list(#{quote_table_name(table_name)})", "SCHEMA").map do |row|
+          new_indexes_from_fields(index_definitions(table_name), table_name)
+        end
+
+        def new_indexes_from_fields(fields, table_name)
+          fields.map do |row|
             # Indexes SQLite creates implicitly for internal use start with "sqlite_".
             # See https://www.sqlite.org/fileformat2.html#intschema
             next if row["name"].starts_with?("sqlite_")
```

Reply:
👍 I mainly wanted to add this next to the `def new_column_from_field` method. Happy to move it up though !